In [1]:
import numpy as np

TASK 1

Please find the task 1 completed on the PDF file

TASK 2

1. The optimal Bellman condition for the states FACEBOOK, CLASS1, CLASS2 and CLASS3:

    V*FB = max((-1 + 0.9 * V*FB), (-1 + 0.9 * V*C1))
    
    V*C1 = max((-1 + 0.9 * V*FB), (-2 + 0.9 * V*C2))
    
    V*C2 = max((-2 + 0.9 * V*C3), 0)
    
    V*C3 = max((1 + 0.9 * (0.2 * V*C1 + 0.4 * V*C2 + 0.4 * V*C3)), 10)

2. Use numpy to find the optimal policy using value iteration. 
   Show all the state values at each iteration.

In [2]:
Vs = np.zeros(4)
vnew = np.zeros(4)
pi = np.zeros(4)
theta = 0.00001
i = 0

while True:
    i = 1+i
    vnew[0] = max((-1+0.9*Vs[0]), (-1+0.9*Vs[1]))
    pi[0] = np.argmax([(-1+0.9*Vs[0]), (-1+0.9*Vs[1])])
    d = abs(vnew[0]-Vs[0])
    
    vnew[1] = max((-1+0.9*Vs[0]), (-2+0.9*Vs[2]))
    pi[1] = np.argmax([(-1+0.9*Vs[0]), -float('inf'), (-2+0.9*Vs[2])])
    d = max(d, abs(vnew[1]-Vs[1]))
    
    vnew[2] = max((-2+0.9*Vs[3]), 0)
    pi[2] = np.argmax([-float('inf'), -float('inf'), -float('inf'), (-2+0.9*Vs[2])])
    d = max(d, abs(vnew[2]-Vs[2]))
    
    vnew[3] = max((1+0.9*(0.2*Vs[1]+0.4*Vs[2]+0.4*Vs[3])), 10)
    pi[3] = np.argmax([-float('inf'), -float('inf'), -float('inf'), (1+0.9*(0.2*Vs[1]+0.4*Vs[2]+0.4*Vs[3])), 10])
    d = max(d, abs(vnew[3]-Vs[3]))
    
    print('THE V VALUES AND POLICY VALUES AT ITERATION', i, ':')
    print(vnew)
    print(pi,'\n')

    if (d < theta):
        break
        
    Vs[0] = vnew[0]
    Vs[1] = vnew[1]
    Vs[2] = vnew[2] 
    Vs[3] = vnew[3]
    

THE V VALUES AND POLICY VALUES AT ITERATION 1 :
[-1. -1.  0. 10.]
[0. 0. 3. 4.] 

THE V VALUES AND POLICY VALUES AT ITERATION 2 :
[-1.9 -1.9  7.  10. ]
[0. 0. 3. 4.] 

THE V VALUES AND POLICY VALUES AT ITERATION 3 :
[-2.71  4.3   7.   10.  ]
[0. 2. 3. 4.] 

THE V VALUES AND POLICY VALUES AT ITERATION 4 :
[ 2.87  4.3   7.   10.  ]
[1. 2. 3. 4.] 

THE V VALUES AND POLICY VALUES AT ITERATION 5 :
[ 2.87  4.3   7.   10.  ]
[1. 2. 3. 4.] 



The final V values are:

    V_FB = 2.87

    V_C1 = 4.3

    V_C2 = 7

    V_C3 = 10


The best Policy acquired is:
    
    pi_FB = quit

    pi_C1 = study

    pi_C2 = study

    pi_C3 = sleep
    
    

TASK 3

In [3]:
Rewards = np.array([[-1, -1, -float('inf'), -float('inf'), float('inf')],
                    [-1, -float('inf'), -2, -float('inf'), -float('inf')],
                    [-float('inf'), -float('inf'), -2, -float('inf'), 0],
                    [-float('inf'), -float('inf'), 10, 1, -float('inf')],
                   ])
Action = np.array([3, 3, 2, 3, 2, 4])
State = np.array([3, 3, 2, 3, 1, 2, 4])

In [4]:
def update_q(Qs, state, next_state, action):
    rsa = Rewards[state, action]
    qsa = Qs[state, action]
    new_q = rsa + 0.9 * np.nanmax(Qs[next_state, :])
    print('THE NEW Q VALUE IS: ', new_q)
    Qs[state, action] = float(str(round(new_q, 2)))
    return Qs

In [5]:
Q = np.array([[-1, -1, np.nan, np.nan, np.nan],
             [-1.9, np.nan, -2, np.nan, np.nan],
             [np.nan, np.nan, -1.1, np.nan, 0],
             [np.nan, np.nan, 10, 1, np.nan],
             [0, 0, 0, 0, 0]])

for i in range (len(Action)):
    state = State[i]
    action = Action[i]
    next_state = State[i+1]
    qnew = update_q(Q, state, next_state, action)
    print('Q VALUES AFTER ACTION', i+1, ':')
    print(qnew[0:4,:],'\n')

THE NEW Q VALUE IS:  10.0
Q VALUES AFTER ACTION 1 :
[[-1.  -1.   nan  nan  nan]
 [-1.9  nan -2.   nan  nan]
 [ nan  nan -1.1  nan  0. ]
 [ nan  nan 10.  10.   nan]] 

THE NEW Q VALUE IS:  1.0
Q VALUES AFTER ACTION 2 :
[[-1.  -1.   nan  nan  nan]
 [-1.9  nan -2.   nan  nan]
 [ nan  nan -1.1  nan  0. ]
 [ nan  nan 10.   1.   nan]] 

THE NEW Q VALUE IS:  7.0
Q VALUES AFTER ACTION 3 :
[[-1.  -1.   nan  nan  nan]
 [-1.9  nan -2.   nan  nan]
 [ nan  nan  7.   nan  0. ]
 [ nan  nan 10.   1.   nan]] 

THE NEW Q VALUE IS:  -0.71
Q VALUES AFTER ACTION 4 :
[[-1.   -1.     nan   nan   nan]
 [-1.9    nan -2.     nan   nan]
 [  nan   nan  7.     nan  0.  ]
 [  nan   nan 10.   -0.71   nan]] 

THE NEW Q VALUE IS:  4.3
Q VALUES AFTER ACTION 5 :
[[-1.   -1.     nan   nan   nan]
 [-1.9    nan  4.3    nan   nan]
 [  nan   nan  7.     nan  0.  ]
 [  nan   nan 10.   -0.71   nan]] 

THE NEW Q VALUE IS:  0.0
Q VALUES AFTER ACTION 6 :
[[-1.   -1.     nan   nan   nan]
 [-1.9    nan  4.3    nan   nan]
 [  nan   

The final Q values are:
      
                  FB  quit  study  pub  sleep
        
     Facebook     -1    -1     -     -     -
 
     Class 1     -1.9    -    4.3    -     -

     Class 2       -     -     7     -     0
 
     Class 3       -     -    10   -0.71   -
 
 For complete steps please check the PDF file.